node/libs/concurrency/src/sync/prunable_mpsc/mod.rs

```diff
@@ -0,0 +1,114 @@
+//! Prunable, multi-producer, single-consumer, unbounded FIFO queue for communicating between asynchronous tasks.
+//! The pruning takes place whenever a new value is sent, based on a specified predicate.
+//!
+//! The separation of [`Sender`] and [`Receiver`] is employed primarily because [`Receiver`] requires
+//! a mutable reference to the signaling channel, unlike [`Sender`], hence making it undesirable to
+//! be used in conjunction.
+//!
+use crate::{
+    ctx,
+    sync::{self, watch},
+};
+use std::{
+    collections::VecDeque,
+    fmt,
+    sync::{Arc, Mutex},
+};
+
+#[cfg(test)]
+mod tests;
+
+/// Creates a channel and returns the [`Sender`] and [`Receiver`] handles.
+/// All values sent on [`Sender`] will become available on [`Receiver`] in the same order as it was sent,
+/// unless will be pruned before received.
+/// The Sender can be cloned to send to the same channel from multiple code locations. Only one Receiver is supported.
+///
+/// * [`T`]: The type of data that will be sent through the channel.
+/// * [`pruning_predicate`]: A function that determines whether an unreceived, pending value in the buffer should be pruned based on a newly sent value.
+///
+pub fn channel<T>(
+    pruning_predicate: impl 'static + Sync + Send + Fn(&T, &T) -> bool,
```

Comment:
it still isn't documented what is the semantics of the predicate: if predicate(a,b) is called and it returns true/false which element will be pruned? a or b?

Reply:
Added clarification here: https://github.com/matter-labs/era-consensus/pull/56/commits/8e4543cb1d2963472b466226d977c6972f9dbfec